rpc/src/service_builder.rs

```diff
@@ -201,15 +204,51 @@ impl<'a> ServiceBuilder<'a> {
         db_config: &DBConfig,
         indexer_config: &IndexerConfig,
     ) -> Self {
-        let indexer = IndexerService::new(db_config, indexer_config, shared.async_handle().clone());
+        // Initialize instances of data sources that will be shared for use by indexer and indexer-r.
+        let ckb_secondary_db = new_secondary_db(db_config, &indexer_config.into());
+        let mut pool_service =
+            PoolService::new(indexer_config.index_tx_pool, shared.async_handle().clone());
+
+        // Init indexer service.
+        let indexer = IndexerService::new(
+            ckb_secondary_db.clone(),
+            pool_service.clone(),
+            indexer_config,
+            shared.async_handle().clone(),
+        );
         let indexer_handle = indexer.handle();
-        let rpc_methods = IndexerRpcImpl::new(indexer_handle).to_delegate();
+        let indexer_rpc_methods = IndexerRpcImpl::new(indexer_handle).to_delegate();
+
         if self.config.indexer_enable() {
-            start_indexer(&shared, indexer, indexer_config.index_tx_pool);
-            self.add_methods(rpc_methods);
+            indexer.spawn_poll(shared.notify_controller().clone());
+            if indexer_config.index_tx_pool {
+                pool_service.index_tx_pool(shared.notify_controller().clone());
+            }
+            self.add_methods(indexer_rpc_methods);
+        } else {
+            self.update_disabled_methods("Indexer", indexer_rpc_methods);
+        }
+
+        // Init indexer-r service
+        let indexer_r = IndexerRService::new(
+            ckb_secondary_db,
+            pool_service.clone(),
+            indexer_config,
+            shared.async_handle().clone(),
+        );
+        let indexer_r_handle = indexer_r.handle();
+        let indexer_r_rpc_methods = IndexerRRpcImpl::new(indexer_r_handle).to_delegate();
+
+        if self.config.indexer_r_enable() {
```

Comment:
`indexer_r` seems too short to distinguish from `indexer`, maybe `db_indexer`?

Reply:
Renamed rich indexer.